xgcm/autogenerate.py

```diff
@@ -0,0 +1,254 @@
+                    fill_value = z_pad
+
+            # Input coordinate has to be declared as center,
+            # or xgcm.Axis throws error. Will be rewrapped below.
+            attrs = ds[old_name].attrs
+            attrs['axis'] = ki
+            attrs.pop('c_grid_axis_shift', None)
+            ds[old_name].attrs = attrs
+
+            if dim_switch:
+                ax = AutoGenerate(ds, ki, periodic=periodic, wrap=wrap)
+                ds.coords[new_name] = ax._generate_coord(ds[old_name], to,
+                                                         boundary=boundary,
+                                                         fill_value=fill_value)
+
+            else:
+                ax = Axis(ds, ki, periodic=periodic, wrap=wrap)
+                ds.coords[new_name] = ax.interp(ds[old_name], boundary=boundary,
+                                         fill_value=fill_value)
+
+            attrs_new = ds[new_name].attrs
+            attrs_new['axis'] = ki
+            attrs_new.pop('c_grid_axis_shift', None)
+            # Reset appropriate attributes for old and new coordinates
+            if abs(to_shift) > 0.0:
+                attrs_new['c_grid_axis_shift'] = to_shift
+            ds[new_name].attrs = attrs_new
+
+            if abs(from_shift) > 0.0:
+                ds[old_name].attrs['c_grid_axis_shift'] = from_shift
+
+        return ds
+
+
+def autogenerate_ds(ds,
+                    axes={'X': 'lon', 'Y': 'lat'},
+                    position='center',
+                    coord_axes=None,
+                    x_wrap=360,
+                    y_wrap=180,
+                    z_pad=None,
+                    x_coord_wrap=360,
+                    y_coord_wrap=180,
+                    z_coord_pad=None):
+    """
+    Regenerate all c-grid information from an existing dataset
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        The data on which to operate
+    axes : axes on which to regenerate dims (and optional coords, see
+        'coord_axes' input). Needs to be in the form of a dict with keys
+        ('X','Y'or'Z'), the values need to be the corresponding dimension names
+        in ds.
+    position : {'center', 'left', 'right'}
+        The position of the datapoints relative to the coordinates given
+        This will usually be the cell center, but can sometimes be the
+        lower left (option:'left') or perhaps upper right (option:'right')
+        corner of the gridcell.
+    coord_axes : (optional) Like 'axes' but for multidimensional coordinates in
+        ds
+    x_wrap : float - Value of discontinuity at the 'X' boundary
+    y_wrap : float - Value of discontinuity at the 'Y' boundary
+    z_pad : float -  Depth padding value. If None, it will be determined from
+        data
+    x_coord_wrap : float - Like 'x_wrap' but for coordinate variable
+    y_coord_wrap : float - Like 'y_wrap' but for coordinate variable
+    z_coord_pad : float -  Like 'z_pad' but for coordinate variable
```

Comment:
Again, I am opposed to hard-coding the possible axes. The design should be more general and less redundant.